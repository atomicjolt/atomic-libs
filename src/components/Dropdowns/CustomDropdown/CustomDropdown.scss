@use "../../mixins";
@use "../common.scss";
