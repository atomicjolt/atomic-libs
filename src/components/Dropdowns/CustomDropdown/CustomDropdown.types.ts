import { EventHandler, SharedInputProps } from "../../../types";

export interface CustomDropdownProps<T>
  extends Omit<SharedInputProps, "placeholder"> {
  readonly value: T | null;
  readonly onChange: EventHandler<T, React.MouseEvent<Element>>;
  readonly children: React.ReactElement | React.ReactElement[];
}

export interface CustomSelectContext<T> {
  currentValue: T | null;
  onClick: (value: T, event: React.MouseEvent<Element>) => void;
}
