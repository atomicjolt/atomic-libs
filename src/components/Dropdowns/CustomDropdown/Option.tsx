import React, { useContext } from "react";
import cn from "classnames";
import { Context } from "./utils";

interface CustomSelectOptionProps<T> {
  value: T;
  children: React.ReactNode;
}

/** An Option in the `CustomDropdown` */
export function Option<T>({ value, children }: CustomSelectOptionProps<T>) {
  const { currentValue, onClick } = useContext(Context);

  return (
    <li
      className={cn("aje-combobox__option", {
        "is-focused": value == currentValue,
      })}
      role="option"
      aria-selected={value == currentValue}
      onClick={(e) => onClick(value, e)}
    >
      {children}
    </li>
  );
}

/** A `CustomDropdown` `Option` item with `null` as the value */
export function EmptyOption({ children }: { children?: React.ReactNode }) {
  return <Option value={null}>{children}</Option>;
}
