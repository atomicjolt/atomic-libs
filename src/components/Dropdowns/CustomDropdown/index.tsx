import React, { useContext, useEffect, useMemo, useRef, useState } from "react";
import cn from "classnames";
import Label from "../../Utility/Label";
import { useBool, useClickOutside, useIds } from "../../../hooks";
import { EventHandler, SharedInputProps } from "../../../types";
import InputError from "../../Utility/InputError";

export interface CustomDropdownProps<T>
  extends Omit<SharedInputProps, "placehodler"> {
  value: T | null;
  onChange: EventHandler<T, React.MouseEvent<Element>>;
  children: React.ReactElement | React.ReactElement[];
}

interface CustomSelectContext<T> {
  currentValue: T | null;
  onClick: (value: T, event: React.MouseEvent<Element>) => void;
}

const Context = React.createContext<CustomSelectContext<any>>(
  {} as CustomSelectContext<any>
);

/**
 * Custom Select
 *
 * https://www.w3.org/TR/wai-aria-practices/examples/combobox/combobox-select-only.html for accessibility implementation.
 *
 * */
function CustomDropdown<T>({
  value,
  onChange,
  size = "medium",
  label,
  error,
  message,
  hideLabel = false,
  disabled = false,
  children,
}: CustomDropdownProps<T>) {
  const [menuActive, toggleMenu] = useBool(false);
  const ref = useRef(null);
  const [inputId, listBoxId, errorId, labelId] = useIds("custom-dropdown", [
    "combo",
    "list",
    "errors",
    "label",
  ]);

  const selectedChild = useMemo(() => {
    const childArray = React.Children.toArray(children) as React.ReactElement[];
    childArray.forEach((c) => {
      if (c.type !== CustomDropdown.Option && c.type !== CustomDropdown.Empty) {
        throw Error(
          "All children of CustomSelect must be CustomSelect.Option or CustomSelect.Empty"
        );
      }
    });
    return childArray.find((c) => c.props["value"] == value);
  }, [value]);

  console.log(selectedChild);
  useClickOutside(
    ref,
    () => {
      if (menuActive) toggleMenu();
    },
    { enabled: menuActive }
  );

  const ctx: CustomSelectContext<T> = {
    currentValue: value,
    onClick: onChange,
  };

  return (
    <div
      className={cn("aje-dropdown", `is-${size}`, {
        "has-error": error,
        "is-disabled": disabled,
      })}
    >
      <Label
        message={message}
        htmlFor={inputId}
        id={labelId}
        hidden={hideLabel}
      >
        {label}
      </Label>
      <div className="aje-combobox">
        <div
          className="aje-combobox__input"
          aria-controls={listBoxId}
          aria-expanded={menuActive}
          aria-haspopup="listbox"
          aria-labelledby={labelId}
          aria-describedby={error ? errorId : ""}
          id={inputId}
          role="combobox"
          tabIndex={0}
          onClick={toggleMenu}
        >
          <span>{selectedChild?.props?.children}</span>
        </div>
        <ul
          className="aje-combobox__menu"
          role="listbox"
          id={listBoxId}
          aria-labelledby={labelId}
          tabIndex={-1}
        >
          <Context.Provider value={ctx}>{children}</Context.Provider>
        </ul>
      </div>
      <InputError error={error} id={errorId} />
    </div>
  );
}

interface CustomSelectOptionProps<T> {
  value: T;
  children: React.ReactNode;
}

/** An Option in the `CustomSelect` */
function CustomSelectOption<T>({
  value,
  children,
}: CustomSelectOptionProps<T>) {
  const { currentValue, onClick } = useContext(Context);

  return (
    <li
      className={cn("aje-combobox__option", {
        "is-focused": value == currentValue,
      })}
      role="option"
      aria-selected={value == currentValue}
      onClick={(e) => onClick(value, e)}
    >
      {children}
    </li>
  );
}

CustomDropdown.Option = CustomSelectOption;
// @ts-ignore
CustomDropdown.Option.displayName = "CustomSelect.Option";

/** A `CustomSelect.Option` item with `null` as the value */
function CustomSelectEmpty({ children }: { children?: React.ReactNode }) {
  return <CustomDropdown.Option value={null}>{children}</CustomDropdown.Option>;
}

CustomDropdown.Empty = CustomSelectEmpty;
// @ts-ignore
CustomDropdown.Empty.displayName = "CustomSelect.Empty";

export default CustomDropdown;
