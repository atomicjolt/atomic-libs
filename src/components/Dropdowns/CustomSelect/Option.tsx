import React, { useContext } from "react";

export interface CustomSelectOptionProps<T> {
  value: T;
  children: React.ReactNode;
}

/** An Option in the `CustomSelect` */
export function Option<T>({ value, children }: CustomSelectOptionProps<T>) {
  return null;
}

/** A `CustomSelect` `Option` item with `null` as the value */
export function EmptyOption({ children }: { children?: React.ReactNode }) {
  return <Option value={null}>{children}</Option>;
}
