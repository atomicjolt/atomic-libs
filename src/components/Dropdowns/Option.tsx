import React from "react";

export interface OptionProps<T> {
  value: T;
  children: React.ReactNode;
}

/** An Option in the various Dropdown */
export default function Option<T>({ value, children }: OptionProps<T>) {
  // @ts-ignore
  return <option value={value}>{children}</option>;
}
