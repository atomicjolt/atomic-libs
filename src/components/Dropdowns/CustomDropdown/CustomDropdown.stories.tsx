import { ComponentMeta, ComponentStory } from "@storybook/react";
import CustomDropdown, { Props } from ".";

export default {
title: "Dropdowns/CustomDropdown",
component: CustomDropdown
} as ComponentMeta<typeof CustomDropdown>

const Template: ComponentStory<typeof CustomDropdown> = (args: Props) => (
  <CustomDropdown {...args}>
    <span>Option 1</span>
  </CustomDropdown>
);

export const Default = Template.bind({});
Default.args = {
  label: "Custom dropdown label",
}

// Additional stories go here