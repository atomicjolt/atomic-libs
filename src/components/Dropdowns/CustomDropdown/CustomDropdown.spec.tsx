import React from "react";
import { render } from "@testing-library/react";
import CustomDropdown, { FloatingCustomDropdown } from ".";
import { Option } from "./Option";

describe("matches snapshots", () => {
  const shared = {
    label: "Custom dropdown label",
    value: "opt1",
    onChange: () => {},
  };

  it("matches default variant", () => {
    const result = render(
      <CustomDropdown {...shared}>
        <Option value="opt1">Option 1</Option>
        <Option value="opt2">Option 2</Option>
      </CustomDropdown>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });

  it("matches floating variant", () => {
    const result = render(
      <FloatingCustomDropdown {...shared}>
        <Option value="opt1">Option 1</Option>
        <Option value="opt2">Option 2</Option>
      </FloatingCustomDropdown>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });
});
