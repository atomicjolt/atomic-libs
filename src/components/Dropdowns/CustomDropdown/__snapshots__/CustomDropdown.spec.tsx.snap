// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="aje-dropdown is-medium"
  >
    <label
      class="aje-label"
      for="custom-dropdown-combo-12345678900000000000"
      id="custom-dropdown-label-12345678900000000000"
    >
      Custom dropdown label
    </label>
    <div
      class="aje-combobox"
    >
      <div
        aria-controls="custom-dropdown-list-12345678900000000000"
        aria-describedby=""
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="custom-dropdown-label-12345678900000000000"
        class="aje-combobox__input"
        id="custom-dropdown-combo-12345678900000000000"
        role="combobox"
        tabindex="0"
      >
        <span>
          Option 1
        </span>
      </div>
      <ul
        aria-labelledby="custom-dropdown-label-12345678900000000000"
        class="aje-combobox__menu"
        id="custom-dropdown-list-12345678900000000000"
        role="listbox"
        tabindex="-1"
      >
        <li
          aria-selected="true"
          class="aje-combobox__option is-focused"
          role="option"
        >
          Option 1
        </li>
        <li
          aria-selected="false"
          class="aje-combobox__option"
          role="option"
        >
          Option 2
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
